### PR TITLE
Changed locale retrieval way to a fancy one

### PR DESCRIPTION
**Pull request by toughengineer**
This is a quick and dirty one, but I like it much, much better than #2592.
I hope I haven't screwed up anything.

And this one is **not** slowed down compared to the ugly one with lamdas and whatnot.

### CODE REVIEW

File: include/fmt/chrono.h [hunk 1]
```diff
@@ -961,6 +961,26 @@ OutputIt format_duration_unit(OutputIt out) {
...
+  ~get_locale() {
+    if (has_locale) locale.~locale();
+  }
+  operator const std::locale&() const {
+    return has_locale ? locale : get_classic_locale();
+  }
+
+ private:
+  union {
+    monostate dummy{};
```
Comment (phprus) [inline include/fmt/chrono.h:978]: ``monostate``  is C++17 only

**Comment by toughengineer** [comment]
Yep, I screwed up, but now it seems to be fixed.

**Comment by phprus** [comment]
I suggested PR #2586, which moves all locale-specific formatters into the ``tm_writer`` class.
I have an idea how to simplify the locale-specific selectors (this PR). I can implement it on top of the PR #2586.

**Comment by phprus** [comment]
No, my idea is perf slower :(

**Comment by vitaut** [comment]
LGTM but please rebase.

File: include/fmt/chrono.h (hunk 1 above)
Comment (vitaut) [reply include/fmt/chrono.h:979]: `{}` seems unnecessary, please remove.

Comment (toughengineer) [reply include/fmt/chrono.h:979]: Strictly speaking this is unnecessary.

But without `monospace` default initialized, this is kinda scary because you have a generally unintialized member, and invariants become kinda hard to follow.

All this is OK for experts, but I imagine that this is harder to follow for non-experts.
`monostate` has negligible overhead, but helps to figure out what's happening more easily.
That's why I added `monostate` member.

Comment (vitaut) [reply include/fmt/chrono.h:979]: `monostate` is fine but I don't think you need to initialize it explicitly at least because it has no state =).

Comment (toughengineer) [reply include/fmt/chrono.h]: If you never initialize `monostate dummy`, than there is no point in having it at all.

IMHO, initializing `monostate dummy` by default gives some understanding and peace of mind to the person trying to understand the code.

Initializing it by default says that the lifetime of the `dummy` object is started, and you need to explicitly do something to change it.

In this case when `locale` is initialized using placement `new`, it explicitly says that the lifetime of `dummy` ends and another object starts to occupy the storage simultaneously starting its lifetime.

If there was no `monostate dummy` it would work perfectly correct, but then a person trying to understand the code must keep in mind that the union is generally _uninitialized_.

**Comment by vitaut** [comment]
Merged, thanks!
